pkg/uio/reader.go

```diff
@@ -46,3 +47,18 @@ func ReaderAtEqual(r1, r2 io.ReaderAt) bool {
 	}
 	return bytes.Equal(c, d) && reflect.DeepEqual(r1err, r2err)
 }
+
+// ReadIntoFile reads all from io.Reader into the file at given path.
+//
+// If the file at given path does not exist, a new file will be created.
+// If the file exists at the given path, but not empty, it will be truncated.
+func ReadIntoFile(r io.Reader, p string) error {
+	f, err := os.OpenFile(p, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	b := make([]byte, 512)
+	_, err = io.CopyBuffer(f, r, b)
```

Comment:
I know chris mentioned copybuffer but why not just io.Copy and skip the slice allocation? Note that in some cases io.CopyBuffer won't use the slice anyway ...

Reply:
"If either src implements WriterTo or dst implements ReaderFrom, buf will not be used to perform the copy."

The buffer is a single-thread, and in your case, non-page-sized resource, and that will limit how much the kernel can improve the data movement.

Reply:
Tl;DR - Certainly could switch to io.Copy, want to understand your point here first:

> The buffer is a single-thread, and in your case, non-page-sized resource, and that will limit how much the kernel can improve the data movement

I don't fully follow this: here, I see 512 bytes < 4KB (a typical page size ?), and it 4kb divides 512 bytes well.  While the default buffer size when calling `io.Copy` is 32KB ([32*1024](https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/io/io.go;l=412;drc=refs%2Ftags%2Fgo1.16.7)). I see a page size divides 512 bytes well, so it can fit continuously into pages ? while latter 32KB can continuously occupy whole pages. I don't understand the point of `non-page-sized resource, and that will limit how much the kernel can improve the data movement.`.


`io.Copy` and `io.CopyBuffer` call [copyBuffer](https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/io/io.go;l=381-450), and `WriterTo` or `ReaderFrom` takes precedence over reading via buffered chunks from `Read` interface.

------------
Selected preallocation of 512 for an easily-arguable reason, which I previously replied here https://github.com/u-root/u-root/pull/2087#discussion_r686342019 (marking it resolved may have made it less visible...my bad)

Reply:
If you have a tight for loop that is moving data 512 bytes at a time, the kernel will need to copy that data in, which in turn requires kernel allocations. Further, there is less data moved to amortize the system call overhead.

The Unix contract requires that you not touch the data until the write call returns, but if the data is large and page aligned there are many ways to achieve that goal. The kernel can mark the pages as copy on write, for example. 
There are many efficiencies possible with large, page-aligned IOs that are not available when you use that 512 byte buffer. That is I think one reason that CopyBuffer ignores the buffer in some cases.

So here's an example:
rminnich@a300:/tmp$ >out
rminnich@a300:/tmp$ time ./a 1

real	0m3.702s
user	0m0.553s
sys	0m3.216s
rminnich@a300:/tmp$ >out
rminnich@a300:/tmp$ time ./a

real	0m0.535s
user	0m0.021s
sys	0m0.535s
rminnich@a300:/tmp$ cat a.go
package main

import (
	"io"
	"log"
	"os"
)

func main() {
	f, err := os.Open("in")
	if err != nil {
		log.Fatal(err)
	}
	o, err := os.OpenFile("out", os.O_WRONLY, 0666)
	if err != nil {
		log.Fatal(err)
	}
	b := make([]byte, 512, 512)
	if len(os.Args) > 1 {
		if _, err := io.CopyBuffer(io.Writer(o), io.Reader(f), b); err != nil {
			log.Fatal(err)
		}
	} else {
		if _, err := io.CopyBuffer(o, f, b); err != nil {
			log.Fatal(err)
		}
	}
}
rminnich@a300:/tmp$ 

note the only difference: the first, slow one, uses the buffer. The second one, where I don't force the type, is 8x faster as it ignores the buffer and does larger IOs. The page alignment may not matter in this case, but who knows -- tracing the kernel IO path is not easy.

But maybe I'm missing your point.

Reply:
I am gonna remove the buffer line, and use `io.Copy`.  I still think the example above is irrelevant to the buffer line, but following your leads, I read into implementations of ReadFrom, and `copy_file_range(2)`, and that were definitely fascinating read and learnings 👍  💯 

-----------------------------

I see your point, I see File implements `ReadFrom`

My point was that,  `WriteTo` and `ReadFrom` takes precedence even when buf is passed in. That is contingent on if the Writer or Reader implement them, here it was demonstrated by forcing v.s. no forcing. the type. So the extra line of assigning a buffer of 512 bytes is not relevant much.

The speed advantage here is, I think, from `File`'s ReadFrom implementation, based on system call [copy_file_range(2)](https://man7.org/linux/man-pages/man2/copy_file_range.2.html), which does in-kernel file copy between 2 file descriptors.

From there, if the filesystem where the program were run under implemented copy-on-write, it also adds up speed, given we don't don't any write to the target copies in the process.